Video chapters extraction

kompoth opened the pull request:
Adds property `chapters` to class `YouTube`. Usage example:
```python
from pytube import YouTube 
from datetime import timedelta

yt = YouTube("https://www.youtube.com/watch?v=pvkTC2xIbeY")

for chapter in yt.chapters:
    fancy_start_tm = timedelta(seconds=chapter["timestamp"] / 1000.)
    print(f"{fancy_start_tm} {chapter['label']}"
```
Output:
```
0:00:00 Intro
0:00:10 YouTube Chapters
0:00:45 What are YouTube chapters?
0:01:17 How to add chapter markers?
0:02:42 When are changes updated?
0:03:02 How to disable chapters?
0:03:21 Are chapters available in my country?
```

resolves #1539

kompoth committed:
video chapters extraction

Ripax commented on pytube/extract.py:
```diff
@@ -8,7 +8,8 @@
 
 from pytube.cipher import Cipher
-from pytube.exceptions import HTMLParseError, LiveStreamError, RegexMatchError
+from pytube.exceptions import HTMLParseError, LiveStreamError, \
```
When importing multiple exceptions from a module like pytube.exceptions, it's often considered more readable and maintainable to import them on separate lines rather than using backslashes for line continuation. Here's a cleaner way to import these exceptions
`from pytube.exceptions import (
    HTMLParseError,
    LiveStreamError,
    RegexMatchError,
    ExtractError
)
`

kompoth replied:
Hi, sorry for a delay! Fixed import style in the second commit.

CJNA replied:
This is done @Ripax PTAL.

JuanBindez commented:
hello contribute to this, I'll be merging more often ->  https://github.com/JuanBindez/pytubefix

kompoth committed:
import style fix

CrossNox commented:
Hi! Is anything blocking this to be merged?

CJNA commented:
Bump on this to merge. Until this is merged, will use this local branch

CJNA reviewed (approved):
LGTM

CJNA commented:
Please label this as #enhancement
'Suggest your change as an issue in the with the label #enhancement'

For now pinging maintainers.
@nficano @hbmartin @tfdahlin @swiftyy-mage

CJNA commented:
@RONNCC Seems like a maintainer. Sorry for wrong pings

kompoth commented:
Hi @CJNA. I don't see any way to add a label to the PR. Should I create a linked issue with this label? Thanks.

RONNCC commented:
Hey if you could add a PR with unit tests I'd be happy to review (just tag
me)

On Thu, Feb 29, 2024 at 10:34 PM Vasily Negrebetskiy <
***@***.***> wrote:

> Hi @CJNA <https://github.com/CJNA>. I don't see any way to add a label to
> the PR. Should I create a linked issue with this label? Thanks.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pytube/pytube/pull/1780#issuecomment-1972595644>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKAWHQ3PEHQTBUBXQH5YJ3YWAORDAVCNFSM6AAAAAA4BZYS76VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZSGU4TKNRUGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kompoth committed:
Chaptered video playback

kompoth committed:
test and error handling

kompoth committed:
renamed test

kompoth commented:
@RONNCC hi, I have added a unit test asserting chapters from a given video url. Unfortunately, I didn't manage to use cached json.gzip data for this video. Actually, all unit tests using cached video objects from `test/conftest.py` fail on my environment with the following traceback:
```
    @pytest.fixture
    def cipher_signature():
        """Youtube instance initialized with video id 2lAe1cqCOXo."""
        filename = "yt-video-2lAe1cqCOXo-html.json.gz"
>       return load_and_init_from_playback_file(filename)

tests/conftest.py:51: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/lib/python3.11/unittest/mock.py:1375: in patched
    return func(*newargs, **newkeywargs)
tests/conftest.py:43: in load_and_init_from_playback_file
    v.fmt_streams
pytube/__main__.py:176: in fmt_streams
    stream_manifest = extract.apply_descrambler(self.streaming_data)
pytube/__main__.py:160: in streaming_data
    self.bypass_age_gate()
pytube/__main__.py:257: in bypass_age_gate
    innertube_response = innertube.player(self.video_id)
pytube/innertube.py:455: in player
    return self._call_api(endpoint, query, self.base_data)
pytube/innertube.py:396: in _call_api
    return json.loads(response.read())
/usr/lib/python3.11/unittest/mock.py:1124: in __call__
    return self._mock_call(*args, **kwargs)
/usr/lib/python3.11/unittest/mock.py:1128: in _mock_call
    return self._execute_mock_call(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <Mock name='urlopen().read' id='124152198383376'>, args = (), kwargs = {}, effect = <list_iterator object at 0x70ea6f912800>

    def _execute_mock_call(self, /, *args, **kwargs):
        # separate from _increment_mock_call so that awaited functions are
        # executed separately from their call, also AsyncMock overrides this method
    
        effect = self.side_effect
        if effect is not None:
            if _is_exception(effect):
                raise effect
            elif not _callable(effect):
>               result = next(effect)
E               StopIteration

/usr/lib/python3.11/unittest/mock.py:1185: StopIteration
```
I run it on Python 3.11.6 and pytest 8.0.2.

Despite the errors, I've added a video playback json.gzip `tests/mocks/yt-video-pvkTC2xIbeY-html.json.gz` to the PR and implemented its loader in the `tests/conftest.py`.

Currently the unit test looks like this, but I understand it must use playbacks to be accepted (right?):
```
def test_chapters(
    # video_with_chapters
):
    # TODO: use a preloaded playback
    expected = [
        {'timestamp': 0, 'label': 'Intro'},
        {'timestamp': 10000, 'label': 'YouTube Chapters'},
        {'timestamp': 45000, 'label': 'What are YouTube chapters?'},
        {'timestamp': 77000, 'label': 'How to add chapter markers?'},
        {'timestamp': 162000, 'label': 'When are changes updated?'},
        {'timestamp': 182000, 'label': 'How to disable chapters?'},
        {'timestamp': 201000, 'label': 'Are chapters available in my country?'}
    ]
    # assert video_with_chapters.chapters == expected
    youtube = YouTube('https://www.youtube.com/watch?v=pvkTC2xIbeY')
    assert youtube.chapters == expected
```
I execute it like this: `pytest tests/test_main.py -k 'test_chapters'`.

Could you please verify that I understand everything correctly? Also, it would be great if you have any clues what am I doing wrong with playbacks.

RONNCC commented:
Hmm

On Mon, Mar 4, 2024, 11:47 PM Vasily Negrebetskiy ***@***.***>
wrote:

> @RONNCC <https://github.com/RONNCC> hi, I have added a unit test
> asserting chapters from a given video url. Unfortunately, I didn't manage
> to use cached json.gzip data for this video. Actually, all unit tests using
> cached video objects from test/conftest.py fail on my environment with
> the following traceback:
>
>     @pytest.fixture
>     def cipher_signature():
>         """Youtube instance initialized with video id 2lAe1cqCOXo."""
>         filename = "yt-video-2lAe1cqCOXo-html.json.gz"
> >       return load_and_init_from_playback_file(filename)
>
> tests/conftest.py:51:
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> /usr/lib/python3.11/unittest/mock.py:1375: in patched
>     return func(*newargs, **newkeywargs)
> tests/conftest.py:43: in load_and_init_from_playback_file
>     v.fmt_streams
> pytube/__main__.py:176: in fmt_streams
>     stream_manifest = extract.apply_descrambler(self.streaming_data)
> pytube/__main__.py:160: in streaming_data
>     self.bypass_age_gate()
> pytube/__main__.py:257: in bypass_age_gate
>     innertube_response = innertube.player(self.video_id)
> pytube/innertube.py:455: in player
>     return self._call_api(endpoint, query, self.base_data)
> pytube/innertube.py:396: in _call_api
>     return json.loads(response.read())
> /usr/lib/python3.11/unittest/mock.py:1124: in __call__
>     return self._mock_call(*args, **kwargs)
> /usr/lib/python3.11/unittest/mock.py:1128: in _mock_call
>     return self._execute_mock_call(*args, **kwargs)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
>
> self = <Mock name='urlopen().read' id='124152198383376'>, args = (), kwargs = {}, effect = <list_iterator object at 0x70ea6f912800>
>
>     def _execute_mock_call(self, /, *args, **kwargs):
>         # separate from _increment_mock_call so that awaited functions are
>         # executed separately from their call, also AsyncMock overrides this method
>
>         effect = self.side_effect
>         if effect is not None:
>             if _is_exception(effect):
>                 raise effect
>             elif not _callable(effect):
> >               result = next(effect)
> E               StopIteration
>
> /usr/lib/python3.11/unittest/mock.py:1185: StopIteration
>
> I run it on Python 3.11.6 and pytest 8.0.2.
>
> Despite the errors, I've added a video playback json.gzip
> tests/mocks/yt-video-pvkTC2xIbeY-html.json.gz to the PR and implemented
> its loader in the tests/conftest.py.
>
> Currently the unit test looks like this, but I understand it must use
> playbacks to be accepted (right?):
>
> def test_chapters(
>     # video_with_chapters
> ):
>     # TODO: use a preloaded playback
>     expected = [
>         {'timestamp': 0, 'label': 'Intro'},
>         {'timestamp': 10000, 'label': 'YouTube Chapters'},
>         {'timestamp': 45000, 'label': 'What are YouTube chapters?'},
>         {'timestamp': 77000, 'label': 'How to add chapter markers?'},
>         {'timestamp': 162000, 'label': 'When are changes updated?'},
>         {'timestamp': 182000, 'label': 'How to disable chapters?'},
>         {'timestamp': 201000, 'label': 'Are chapters available in my country?'}
>     ]
>     # assert video_with_chapters.chapters == expected
>     youtube = YouTube('https://www.youtube.com/watch?v=pvkTC2xIbeY')
>     assert youtube.chapters == expected
>
> I execute it like this: pytest tests/test_main.py -k 'test_chapters'.
>
> Could you please verify that I understand everything correctly? Also, it
> would be great if you have any clues what am I doing wrong with playbacks.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pytube/pytube/pull/1780#issuecomment-1978142854>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKAWHXF74I2SMY3BXN2KKTYWV2AFAVCNFSM6AAAAAA4BZYS76VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZYGE2DEOBVGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kompoth commented:
@RONNCC @Ripax I am already a bit out of context, but AFAIR it was almost ready. Could you please advise?
